File: src/analysis/assemble-briefing-data.mjs
#!/usr/bin/env node
// Deterministic briefing data assembler.
// Consolidates all pipeline analysis outputs into a single briefing-data.json.
// Same inputs always produce byte-identical output.
//
// Usage: node assemble-briefing-data.mjs --dir <output/YYYY-MM-DD_slug/>
// Outputs briefing-data.json to the same directory.

import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { join, basename } from 'node:path';

const PIPELINE_VERSION = '0.2.0';

// --- CLI parsing ---
const args = process.argv.slice(2);
function flag(name) {
  const idx = args.indexOf(name);
  if (idx === -1 || idx + 1 >= args.length) return undefined;
  return args[idx + 1];
}

const dir = flag('--dir');

if (dir === undefined) {
  console.error('Usage: node assemble-briefing-data.mjs --dir <output/YYYY-MM-DD_slug/>');
  process.exit(1);
}

// --- File discovery ---
const INPUT_FILES = {
  serp: 'serp-processed.json',
  keywordsProcessed: 'keywords-processed.json',
  keywordsFiltered: 'keywords-filtered.json',
  pageStructure: 'page-structure.json',
  contentTopics: 'content-topics.json',
  entityProminence: 'entity-prominence.json',
  competitorsData: 'competitors-data.json',
};

function loadOptional(filename) {
  const path = join(dir, filename);
  if (existsSync(path) === false) return null;
  try {
    return JSON.parse(readFileSync(path, 'utf-8'));
  } catch {
    return null;
  }
}

const data = {};
for (const [key, filename] of Object.entries(INPUT_FILES)) {
  data[key] = loadOptional(filename);
}

// --- Extract meta ---
// Date from directory name pattern YYYY-MM-DD_slug, fallback to today
function extractDateFromDir(dirPath) {
  const dirName = basename(dirPath);
  const match = dirName.match(/^(\d{4}-\d{2}-\d{2})_/);
  if (match) return match[1];
  return new Date().toISOString().slice(0, 10);
}

const dateStr = extractDateFromDir(dir);
const currentYear = parseInt(dateStr.slice(0, 4), 10);

// Seed keyword from keywords-processed or keywords-filtered or serp
const seedKeyword = data.keywordsProcessed?.seed_keyword
  || data.keywordsFiltered?.seed_keyword
  || data.serp?.target_keyword
  || null;

// --- Year normalization ---
// Replace 2024/2025 with current year in string values (recursive)
function normalizeYears(value) {
  if (value === null || value === undefined) return value;
  if (typeof value === 'string') {
    return value.replace(/\b(2024|2025)\b/g, String(currentYear));
  }
  if (Array.isArray(value)) {
    return value.map(v => normalizeYears(v));
  }
  if (typeof value === 'object') {
    const result = {};
    const keys = Object.keys(value).sort();
    for (const k of keys) {
      result[k] = normalizeYears(value[k]);
    }
    return result;
  }
  return value;
}

// --- 1. Cluster ranking: sort keyword clusters by total search volume desc ---
function buildClusterRanking() {
  const source = data.keywordsFiltered || data.keywordsProcessed;
  if (source === null || Array.isArray(source.clusters) === false) return [];

  const ranked = source.clusters.map(cluster => {
    const keywords = Array.isArray(cluster.keywords) ? cluster.keywords : [];
    const totalVolume = keywords.reduce((sum, kw) => sum + (kw.search_volume || 0), 0);
    return {
      cluster_keyword: cluster.cluster_keyword,
      cluster_label: cluster.cluster_label || null,
      keyword_count: cluster.keyword_count || keywords.length,
      total_search_volume: totalVolume,
      cluster_opportunity: cluster.cluster_opportunity ?? null,
      rank: 0, // placeholder, assigned after sort
    };
  });

  // Sort by total_search_volume desc, then cluster_keyword alphabetically for determinism
  ranked.sort((a, b) => {
    if (b.total_search_volume !== a.total_search_volume) {
      return b.total_search_volume - a.total_search_volume;
    }
    return a.cluster_keyword.localeCompare(b.cluster_keyword);
  });

  // Assign rank
  for (let i = 0; i < ranked.length; i++) {
    ranked[i].rank = i + 1;
  }

  return ranked;
}

// --- 2. Proof keywords from content-topics ---
function buildProofKeywords() {
  if (data.contentTopics === null) return null;
  return normalizeYears(data.contentTopics.proof_keywords || []);
}

// --- 3. Module frequency from page-structure ---
function buildModuleFrequency() {
  if (data.pageStructure === null) return null;
  const cc = data.pageStructure.cross_competitor;
  if (cc === undefined || cc === null) return null;
  return {
    common_modules: cc.common_modules || [],
    rare_modules: cc.rare_modules || [],
    module_frequency: cc.module_frequency || {},
  };
}

// --- 4. Section weights from content-topics ---
function buildSectionWeights() {
  if (data.contentTopics === null) return null;
  return data.contentTopics.section_weights || [];
}

// --- 5. AIO data from serp ---
function buildAioData() {
  if (data.serp === null) return null;
  const aio = data.serp.serp_features?.ai_overview;
  if (aio === undefined || aio === null) return { present: false };
  return normalizeYears(aio);
}

// --- 6. FAQ with priority + cluster mapping from filter-keywords ---
function buildFaqData() {
  if (data.keywordsFiltered === null) return null;
  const faqSelection = data.keywordsFiltered.faq_selection;
  if (Array.isArray(faqSelection) === false) return null;

  const questions = normalizeYears(faqSelection);
  // Determine source: PAA questions come from SERP
  return {
    questions,
    paa_source: 'serp',
  };
}

// --- 7. Entity candidates with prominence ---
function buildEntityCandidates() {
  // Base entity candidates from content-topics
  const baseCandidates = data.contentTopics?.entity_candidates || null;
  if (baseCandidates === null) return null;

  // If entity prominence data exists, merge prominence info
  if (data.entityProminence === null) return baseCandidates;

  const prominenceMap = new Map();
  const clusters = data.entityProminence.entity_clusters;
  if (Array.isArray(clusters)) {
    for (const cluster of clusters) {
      if (Array.isArray(cluster.entities)) {
        for (const entity of cluster.entities) {
          prominenceMap.set(entity.entity.toLowerCase(), {
            prominence: entity.prominence,
            prominence_source: entity.prominence_source || null,
          });
        }
      }
    }
  }

  return baseCandidates.map(candidate => {
    const promData = prominenceMap.get(candidate.term.toLowerCase());
    if (promData) {
      return {
        ...candidate,
        prominence: promData.prominence,
        prominence_source: promData.prominence_source,
      };
    }
    return candidate;
  });
}

// --- 8. SERP features ---
function buildSerpFeatures() {
  if (data.serp === null) return null;
  const features = data.serp.serp_features;
  if (features === undefined || features === null) return null;
  // Extract a summary of which features are present
  const summary = {};
  for (const [key, val] of Object.entries(features)) {
    if (key === 'ai_overview') {
      summary[key] = val?.present || false;
    } else if (key === 'featured_snippet') {
      summary[key] = val?.present || false;
    } else if (key === 'knowledge_graph') {
      summary[key] = val?.present || false;
    } else if (Array.isArray(val)) {
      summary[key] = val.length > 0;
    } else if (typeof val === 'object' && val !== null) {
      // For signal objects (commercial_signals, local_signals), check if any are true
      summary[key] = Object.values(val).some(v => v === true);
    } else {
      summary[key] = false;
    }
  }
  return summary;
}

// --- 9. Competitors from competitors-data or serp ---
function buildCompetitors() {
  if (data.competitorsData !== null) {
    return normalizeYears(data.competitorsData.competitors || []);
  }
  if (data.serp !== null) {
    return normalizeYears(data.serp.competitors || []);
  }
  return null;
}

// --- 10. Page structures from page-structure ---
function buildPageStructures() {
  if (data.pageStructure === null) return null;
  return data.pageStructure.competitors || [];
}

// --- 11. Content format signals from content-topics ---
function buildContentFormatSignals() {
  if (data.contentTopics === null) return null;
  return data.contentTopics.content_format_signals || {};
}

// --- 12. Stats summary -- quick-glance pipeline coverage metrics ---
function buildStatsSummary() {
  const source = data.keywordsFiltered || data.keywordsProcessed;
  const totalKeywords = source?.total_keywords ?? 0;
  const filteredCount = data.keywordsFiltered?.filtered_keywords ?? totalKeywords;
  const clusterCount = Array.isArray(source?.clusters) ? source.clusters.length : 0;
  const competitorCount = (data.competitorsData?.competitors || data.serp?.competitors || []).length;

  return {
    total_keywords: totalKeywords,
    filtered_keywords: filteredCount,
    total_clusters: clusterCount,
    competitor_count: competitorCount,
  };
}


// --- Keyword data ---
function buildKeywordData() {
  const clusters = buildClusterRanking();
  const source = data.keywordsFiltered || data.keywordsProcessed;
  return {
    clusters: normalizeYears(clusters),
    total_keywords: source?.total_keywords ?? 0,
    filtered_count: data.keywordsFiltered?.filtered_keywords ?? source?.total_keywords ?? 0,
    removal_summary: data.keywordsFiltered?.removal_summary ?? null,
  };
}

// --- Average word count from page structure ---
function buildAvgWordCount() {
  if (data.pageStructure === null) return null;
  return data.pageStructure.cross_competitor?.avg_word_count ?? null;
}

// --- Assemble final output ---
const output = {
  meta: {
    seed_keyword: seedKeyword,
    date: dateStr,
    current_year: currentYear,
    pipeline_version: PIPELINE_VERSION,
  },
  stats: buildStatsSummary(),
  keyword_data: buildKeywordData(),
  serp_data: {
    competitors: buildCompetitors(),
    serp_features: buildSerpFeatures(),
    aio: buildAioData(),
  },
  content_analysis: {
    proof_keywords: buildProofKeywords(),
    entity_candidates: buildEntityCandidates(),
    section_weights: buildSectionWeights(),
    content_format_signals: buildContentFormatSignals(),
  },
  competitor_analysis: {
    page_structures: buildPageStructures(),
    common_modules: buildModuleFrequency()?.common_modules ?? null,
    rare_modules: buildModuleFrequency()?.rare_modules ?? null,
    avg_word_count: buildAvgWordCount(),
  },
  faq_data: buildFaqData(),
  qualitative: {
    entity_clusters: null,
    unique_angles: null,
    content_format_recommendation: null,
    geo_audit: null,
    aio_strategy: null,
    briefing: null,
  },
};

// Write to file and stdout
const jsonStr = JSON.stringify(output, null, 2) + '\n';
writeFileSync(join(dir, 'briefing-data.json'), jsonStr);
console.log(jsonStr);

File: seo_pipeline/analysis/assemble_briefing_data.py
"""Deterministic briefing data assembler.

Consolidates all pipeline analysis outputs into a single briefing-data.json.
Same inputs always produce byte-identical output.

Usage:
    python -m seo_pipeline.analysis.assemble_briefing_data \
        --dir <output/YYYY-MM-DD_slug/> [--market de] [--language de] \
        [--user-domain example.com] [--business-context "..."] [--output path]
"""

from __future__ import annotations

import argparse
import json
import logging
import re
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from seo_pipeline.utils.math import normalize_number

logger = logging.getLogger(__name__)

PIPELINE_VERSION = "0.2.0"

_YEAR_RE = re.compile(r"\b(2024|2025)\b")

# Maps logical names to filenames on disk.
INPUT_FILES = {
    "serp": "serp-processed.json",
    "serp_raw": "serp-raw.json",
    "keywords_processed": "keywords-processed.json",
    "keywords_filtered": "keywords-filtered.json",
    "page_structure": "page-structure.json",
    "content_topics": "content-topics.json",
    "entity_prominence": "entity-prominence.json",
    "competitors_data": "competitors-data.json",
}


# ---------------------------------------------------------------------------
# File discovery
# ---------------------------------------------------------------------------


def _load_optional(directory: Path, filename: str) -> Any:
    """Load a JSON file from *directory*, returning None on missing/invalid."""
    path = directory / filename
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None


def _load_inputs(directory: Path) -> dict[str, Any]:
    """Load all 8 optional input files from *directory*."""
    return {
        key: _load_optional(directory, filename)
        for key, filename in INPUT_FILES.items()
    }


# ---------------------------------------------------------------------------
# Date extraction
# ---------------------------------------------------------------------------

_DIR_DATE_RE = re.compile(r"^(\d{4}-\d{2}-\d{2})_")


def _extract_date_from_dir(dir_path: Path) -> str:
    """Extract YYYY-MM-DD from directory name, fallback to today."""
    match = _DIR_DATE_RE.match(dir_path.name)
    if match:
        return match.group(1)
    return datetime.now(tz=timezone.utc).strftime("%Y-%m-%d")


# ---------------------------------------------------------------------------
# Year normalization
# ---------------------------------------------------------------------------


def normalize_years(value: Any, current_year: int) -> Any:
    """Replace 2024/2025 with *current_year* in all string values (recursive).

    Dict keys are sorted alphabetically to match the Node.js implementation.
    """
    if value is None:
        return value
    if isinstance(value, str):
        return _YEAR_RE.sub(str(current_year), value)
    if isinstance(value, list):
        return [normalize_years(v, current_year) for v in value]
    if isinstance(value, dict):
        return {
            k: normalize_years(v, current_year)
            for k, v in sorted(value.items())
        }
    return value


# ---------------------------------------------------------------------------
# Section builders
# ---------------------------------------------------------------------------


def _build_cluster_ranking(data: dict[str, Any]) -> list[dict[str, Any]]:
    """Sort keyword clusters by total search volume desc, assign rank."""
    source = data["keywords_filtered"] or data["keywords_processed"]
    if source is None or not isinstance(source.get("clusters"), list):
        return []

    ranked = []
    for cluster in source["clusters"]:
        raw_kw = cluster.get("keywords")
        keywords = raw_kw if isinstance(raw_kw, list) else []
        total_volume = sum(kw.get("search_volume") or 0 for kw in keywords)
        ranked.append({
            "cluster_keyword": cluster["cluster_keyword"],
            "cluster_label": cluster.get("cluster_label"),
            "cluster_opportunity": cluster.get("cluster_opportunity"),
            "keyword_count": cluster.get("keyword_count") or len(keywords),
            "rank": 0,
            "total_search_volume": total_volume,
        })

    # Sort by total_search_volume desc, then cluster_keyword asc for determinism
    ranked.sort(key=lambda c: (-c["total_search_volume"], c["cluster_keyword"]))

    for i, item in enumerate(ranked):
        item["rank"] = i + 1

    return ranked


def _build_proof_keywords(
    data: dict[str, Any], current_year: int,
) -> list[dict[str, Any]] | None:
    if data["content_topics"] is None:
        return None
    pks = data["content_topics"].get("proof_keywords", [])
    return normalize_years(pks, current_year)


def _build_module_frequency(data: dict[str, Any]) -> dict[str, Any] | None:
    if data["page_structure"] is None:
        return None
    cc = data["page_structure"].get("cross_competitor")
    if cc is None:
        return None
    return {
        "common_modules": cc.get("common_modules", []),
        "rare_modules": cc.get("rare_modules", []),
        "module_frequency": cc.get("module_frequency", {}),
    }


def _build_section_weights(data: dict[str, Any]) -> list[dict[str, Any]] | None:
    if data["content_topics"] is None:
        return None
    return data["content_topics"].get("section_weights", [])


def _build_aio_data(data: dict[str, Any], current_year: int) -> dict[str, Any] | None:
    if data["serp"] is None:
        return None
    aio = (data["serp"].get("serp_features") or {}).get("ai_overview")
    if aio is None:
        return {"present": False}
    return normalize_years(aio, current_year)


def _build_faq_data(data: dict[str, Any], current_year: int) -> dict[str, Any] | None:
    if data["keywords_filtered"] is None:
        return None
    faq_selection = data["keywords_filtered"].get("faq_selection")
    if not isinstance(faq_selection, list):
        return None
    return {
        "questions": normalize_years(faq_selection, current_year),
        "paa_source": "serp",
    }


def _build_entity_candidates(data: dict[str, Any]) -> list[dict[str, Any]] | None:
    base_candidates = (data["content_topics"] or {}).get("entity_candidates")
    if base_candidates is None:
        return None

    if data["entity_prominence"] is None:
        return base_candidates

    # Build prominence lookup from entity clusters
    prominence_map: dict[str, dict[str, Any]] = {}
    clusters = data["entity_prominence"].get("entity_clusters")
    if isinstance(clusters, list):
        for cluster in clusters:
            if isinstance(cluster.get("entities"), list):
                for entity in cluster["entities"]:
                    prominence_map[entity["entity"].lower()] = {
                        "prominence": entity["prominence"],
                        "prominence_source": entity.get("prominence_source"),
                    }

    result = []
    for candidate in base_candidates:
        prom = prominence_map.get(candidate["term"].lower())
        if prom:
            result.append({
                **candidate,
                "prominence": prom["prominence"],
                "prominence_source": prom["prominence_source"],
            })
        else:
            result.append(candidate)
    return result


def _build_serp_features(data: dict[str, Any]) -> dict[str, bool] | None:
    if data["serp"] is None:
        return None
    features = data["serp"].get("serp_features")
    if features is None:
        return None

    summary: dict[str, bool] = {}
    for key, val in features.items():
        if key in ("ai_overview", "featured_snippet", "knowledge_graph"):
            summary[key] = (val or {}).get("present", False)
        elif isinstance(val, list):
            summary[key] = len(val) > 0
        elif isinstance(val, dict):
            # Signal objects (commercial_signals, local_signals): any True value
            summary[key] = any(v is True for v in val.values())
        else:
            summary[key] = False
    return summary


def _build_competitors(
    data: dict[str, Any], current_year: int,
) -> list[dict[str, Any]] | None:
    if data["competitors_data"] is not None:
        comps = data["competitors_data"].get("competitors", [])
        return normalize_years(comps, current_year)
    if data["serp"] is not None:
        comps = data["serp"].get("competitors", [])
        return normalize_years(comps, current_year)
    return None


def _build_page_structures(data: dict[str, Any]) -> list[dict[str, Any]] | None:
    if data["page_structure"] is None:
        return None
    return data["page_structure"].get("competitors", [])


def _build_content_format_signals(data: dict[str, Any]) -> dict[str, Any] | None:
    if data["content_topics"] is None:
        return None
    return data["content_topics"].get("content_format_signals", {})


def _build_stats_summary(data: dict[str, Any]) -> dict[str, int]:
    source = data["keywords_filtered"] or data["keywords_processed"]
    total_keywords = (source or {}).get("total_keywords", 0)
    filtered_count = (
        data["keywords_filtered"] or {}
    ).get("filtered_keywords", total_keywords)
    cluster_count = (
        len(source["clusters"])
        if source and isinstance(source.get("clusters"), list)
        else 0
    )
    competitor_count = len(
        (data["competitors_data"] or {}).get("competitors")
        or (data["serp"] or {}).get("competitors")
        or []
    )
    return {
        "total_keywords": total_keywords,
        "filtered_keywords": filtered_count,
        "total_clusters": cluster_count,
        "competitor_count": competitor_count,
    }


def _build_keyword_data(
    data: dict[str, Any], current_year: int,
) -> dict[str, Any]:
    clusters = _build_cluster_ranking(data)
    source = data["keywords_filtered"] or data["keywords_processed"]
    total_kw = (source or {}).get("total_keywords", 0)
    filtered = (
        (data["keywords_filtered"] or {}).get("filtered_keywords")
        if data["keywords_filtered"] is not None
        else (source or {}).get("total_keywords", 0)
    )
    return {
        "clusters": normalize_years(clusters, current_year),
        "total_keywords": total_kw,
        "filtered_count": filtered if filtered is not None else total_kw,
        "removal_summary": (data["keywords_filtered"] or {}).get("removal_summary")
        if data["keywords_filtered"] is not None
        else None,
    }


def _build_avg_word_count(data: dict[str, Any]) -> int | float | None:
    if data["page_structure"] is None:
        return None
    cc = data["page_structure"].get("cross_competitor")
    if cc is None:
        return None
    return cc.get("avg_word_count")


# ---------------------------------------------------------------------------
# Main assembly
# ---------------------------------------------------------------------------


def assemble_briefing_data(
    directory: Path,
    *,
    market: str | None = None,
    language: str | None = None,
    user_domain: str | None = None,
    business_context: str | None = None,
    timestamp_override: str | None = None,
) -> dict[str, Any]:
    """Assemble all pipeline outputs into a single briefing-data dict.

    Parameters
    ----------
    directory:
        Path to the output directory containing analysis JSON files.
    market, language, user_domain, business_context:
        Optional metadata flags passed through to the meta section.
    timestamp_override:
        If set, used instead of ``datetime.now()`` for ``phase1_completed_at``.
        Useful for deterministic testing.

    Returns
    -------
    dict
        The assembled briefing data structure.
    """
    data = _load_inputs(directory)

    found = [k for k, v in data.items() if v is not None]
    missing = [k for k, v in data.items() if v is None]
    logger.info(
        "Briefing assembly: found %d/%d input files (missing: %s)",
        len(found),
        len(data),
        ", ".join(missing) if missing else "none",
    )

    date_str = _extract_date_from_dir(directory)
    current_year = int(date_str[:4])

    # Seed keyword from keywords-processed, keywords-filtered, or serp
    seed_keyword = (
        (data["keywords_processed"] or {}).get("seed_keyword")
        or (data["keywords_filtered"] or {}).get("seed_keyword")
        or (data["serp"] or {}).get("target_keyword")
    )

    phase1_completed_at = (
        timestamp_override
        if timestamp_override is not None
        else datetime.now(tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.") + "000Z"
    )

    module_freq = _build_module_frequency(data)

    return {
        "meta": {
            "seed_keyword": seed_keyword,
            "date": date_str,
            "current_year": current_year,
            "pipeline_version": PIPELINE_VERSION,
            "market": market,
            "language": language,
            "user_domain": user_domain,
            "business_context": business_context,
            "phase1_completed_at": phase1_completed_at,
            "data_sources": {
                "competitor_urls": [
                    c["url"]
                    for c in (data["serp"] or {}).get("competitors", [])
                ],
                "location_code": (
                    ((data["serp_raw"] or {}).get("tasks") or [{}])[0]
                    .get("data", {})
                    .get("location_code")
                ),
            },
        },
        "stats": _build_stats_summary(data),
        "keyword_data": _build_keyword_data(data, current_year),
        "serp_data": {
            "competitors": _build_competitors(data, current_year),
            "serp_features": _build_serp_features(data),
            "aio": _build_aio_data(data, current_year),
        },
        "content_analysis": {
            "proof_keywords": _build_proof_keywords(data, current_year),
            "entity_candidates": _build_entity_candidates(data),
            "section_weights": _build_section_weights(data),
            "content_format_signals": _build_content_format_signals(data),
        },
        "competitor_analysis": {
            "page_structures": _build_page_structures(data),
            "common_modules": module_freq["common_modules"] if module_freq else None,
            "rare_modules": module_freq["rare_modules"] if module_freq else None,
            "avg_word_count": _build_avg_word_count(data),
        },
        "faq_data": _build_faq_data(data, current_year),
        "qualitative": {
            "entity_clusters": None,
            "unique_angles": None,
            "content_format_recommendation": None,
            "geo_audit": None,
            "aio_strategy": None,
            "briefing": None,
        },
    }


# ---------------------------------------------------------------------------
# Serialization
# ---------------------------------------------------------------------------


def _normalize_tree(obj: Any) -> Any:
    """Recursively apply normalize_number to all numeric values."""
    if isinstance(obj, dict):
        return {k: _normalize_tree(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_normalize_tree(v) for v in obj]
    if isinstance(obj, float):
        return normalize_number(obj)
    return obj


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Assemble briefing data from pipeline outputs.",
    )
    parser.add_argument(
        "--dir", type=Path, required=True,
        help="Output directory containing analysis JSON files.",
    )
    parser.add_argument("--market", default=None, help="Market identifier.")
    parser.add_argument("--language", default=None, help="Language code.")
    parser.add_argument("--user-domain", default=None, help="User domain.")
    parser.add_argument("--business-context", default=None, help="Business context.")
    parser.add_argument(
        "--output", type=Path, default=None,
        help="Output file path (default: <dir>/briefing-data.json).",
    )
    return parser


def main(argv: list[str] | None = None) -> None:
    """CLI entry point."""
    args = _build_parser().parse_args(argv)
    result = assemble_briefing_data(
        args.dir,
        market=args.market,
        language=args.language,
        user_domain=args.user_domain,
        business_context=args.business_context,
    )
    output_dict = _normalize_tree(result)
    output_json = json.dumps(output_dict, indent=2, ensure_ascii=False) + "\n"

    output_path = args.output or (args.dir / "briefing-data.json")
    output_path.write_text(output_json, encoding="utf-8")
    print("Wrote briefing-data.json", file=sys.stderr)


if __name__ == "__main__":
    main()
